<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>
  <script src="./MyPromise.js"></script>
</head>

<body>
  <script>
    // then 方法的测试代码 ------------------------------------------------------------
    // let p = new MyPromise((resolve, reject) => {
    //   // 改变 promise 的状态的三种方法：resolve、reject、throw
    //   // 同步处理程序
    //   // resolve('ok');
    //   // reject('err');
    //   // throw 'error';

    //   // 异步处理程序
    //   setTimeout(() => {
    //     resolve('ok');
    //     // reject('err');
    //   }, 1000);
    // });
    // // console.log(p);
    // // p.then((value) => {
    // //   // console.log(value);
    // // }, (reason) => {
    // //   // 输出警告
    // //   // console.warn(reason);
    // // });
    // let res = p.then((value) => {
    //   // alert(value);
    //   // 1、返回一个普通的值
    //   // return '返回一个普通的值'
    //   ///2、返回一个promise
    //   // return new MyPromise((resolve, reject) => {
    //   //   // resolve('成功了 返回一个promise');
    //   //   reject('失败了 返回一个promise')
    //   // })
    //   ///3、抛出一个错误
    //   throw '抛出一个错误';
    // }, (reason) => {
    //   // 输出警告
    //   // console.warn(reason);
    //   throw '抛出一个错误';
    // });
    // // console.log(p);
    // // 注意：实例中为异步执行程序时，这里的打印结果的状态一开始是pending，等异步执行结束之后才会改为成功或失败，因此在浏览器控制台输出时要等一下再展开这个打印的promise对象，看到的才是正确的状态；
    // console.log(res);


    // catch 方法的测试代码 ------------------------------------------------------------
    // let p = new MyPromise((resolve, reject) => {
    //   resolve(111);
    //   // reject('err');
    //   // throw 'error';

    //   // 异步处理程序
    //   // setTimeout(() => {
    //   //   reject('err');
    //   // }, 1000);
    // });
    // let res = p.then()
    //   .then()
    //   .then(value => {
    //     console.log(111);
    //     throw '抛出一个错误';
    //   })
    //   .then(value => {
    //     console.log(222)
    //   })
    //   .catch((reason) => {
    //     // 输出警告
    //     console.warn(reason);
    //     // throw '抛出一个错误';
    //   });
    // console.log(res);


    // Promsie.resolve 方法的测试代码 --------------------------------------------------------------
    // let p = MyPromise.resolve('ok');
    // let p1 = MyPromise.resolve(new MyPromise((resolve, reject) => {
    //   // resolve('success');
    //   // reject('err');
    //   throw 'err';
    // }));
    // p1.catch(err => {
    //   console.log(err);
    // })
    // let p2 = MyPromise.resolve(MyPromise.resolve('success'));
    // console.log(p2);


    // Promise.reject 方法的测试代码 ---------------------------------------------------------------
    // let p = MyPromise.reject('err');
    // let p2 = MyPromise.reject(new MyPromise((resolve, reject) => {
    //   resolve('ok');
    // }))
    // console.log(p2);


    // Promise.all/race 方法的测试代码 ------------------------------------------------------------------
    let p1 = new MyPromise((resolve, reject) => {
      setTimeout(() => {
        resolve('111');
      }, 1000)
    })
    let p2 = MyPromise.resolve('222');
    let p3 = MyPromise.resolve('333');
    let p5 = MyPromise.reject('333');
    // let res = MyPromise.all([p1, p2, p3]);
    // let res = MyPromise.all([p1, p2, p5]);
    let res = MyPromise.race([p1, p2, p3]);
    // let res = MyPromise.race([p1, p2, p5]);
    console.log(res);
  </script>
</body>

</html>